pkg/controllers/dynakube/version/oneAgentVersionReconciler.go

```diff
@@ -0,0 +1,212 @@
+	"github.com/Dynatrace/dynatrace-operator/pkg/oci/registry"
+	"github.com/Dynatrace/dynatrace-operator/pkg/util/timeprovider"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+type OneAgentImageVersionReconciler struct {
+	apiReader      client.Reader
+	dtClient       dtclient.Client
+	registryClient registry.ImageGetter
+	timeProvider   *timeprovider.Provider
+}
+
+var _ ImageVersionReconciler = &OneAgentImageVersionReconciler{}
+
+// move to versionstatus return versionstatus created provide customimage and oneagentversion as param -> option pattern?
+func initializeOneAgentVersionStatus(dynakube *dynatracev1beta1.DynaKube, now *metav1.Time) status.VersionStatus {
+	result := dynakube.Status.OneAgent.VersionStatus
+	result.LastProbeTimestamp = now
+	switch {
+	case dynakube.CustomOneAgentImage() != "":
+		result.Source = status.CustomImageVersionSource
+	case dynakube.FeaturePublicRegistry():
+		result.Source = status.PublicRegistryVersionSource
+	case dynakube.CustomOneAgentVersion() != "":
+		result.Source = status.CustomVersionVersionSource
+	default:
+		result.Source = status.TenantRegistryVersionSource
+	}
+	return result
+}
+
+// Reconcile initializes state used for reconciliation and updates the version status used by the dynakube
+func (reconciler *OneAgentImageVersionReconciler) Reconcile(ctx context.Context, dynakube *dynatracev1beta1.DynaKube) error {
+	current := initializeOneAgentVersionStatus(dynakube, reconciler.timeProvider.Now())
+	if reconciler.needsUpdate(current, dynakube) {
+		updatedStatus, err := reconciler.run(ctx, current, dynakube)
+		if err != nil {
+			return err
+		}
+		reconciler.SetOneAgentHealthConfig(ctx, dynakube)
+		if err := ValidateOneAgentStatus(updatedStatus, dynakube); err != nil {
+			return err
+		}
+		dynakube.Status.OneAgent.VersionStatus = updatedStatus
+	}
+	return nil
+}
+
+func ValidateOneAgentStatus(updatedStatus status.VersionStatus, dynakube *dynatracev1beta1.DynaKube) error {
+	imageVersion := updatedStatus.Version
+	imageType := updatedStatus.Type
+
+	if imageVersion == "" {
+		return errors.New("build version of OneAgent image not set")
+	}
+
+	if imageType == status.ImmutableImageType {
+		if dynakube.ClassicFullStackMode() {
+			return errors.New("immutable OneAgent image in combination with classicFullStack mode is not possible")
+		} else if dynakube.FeatureDisableReadOnlyOneAgent() {
+			return errors.New("immutable OneAgent image in combination with readOnly OneAgent filesystem is not possible")
+		}
+	}
+	log.Info("OneAgent metadata present, image type and image version validated")
+	return nil
+}
+
+func (reconciler *OneAgentImageVersionReconciler) SetOneAgentHealthConfig(ctx context.Context, dynakube *dynatracev1beta1.DynaKube) {
+	healthConfig, err := GetOneAgentHealthConfig(ctx, reconciler.apiReader, reconciler.registryClient, dynakube, dynakube.OneAgentImage())
```

Comment:
this is more of a sidenote and out of scope probably: To get the `HealthConfig` we need to query the image registry, the "same way" as when you call `updateVersionWithImage`, so it would be nice that we do not make the "same" call twice **if possible**

Reply:
good point ☝️